File: apps/grokking-ml-fsharp-lib/src/GrokkingMlLib/Nb15ImageRecognition.fs
/// 原著ノートブック #15 `Chapter_10_Neural_Networks/Image_recognition.ipynb`。
///
/// MNIST の手書き数字 7 万枚を、3 層のネットワークで 10 クラスに分類する。
/// 原著は 6 万件・10 エポックで正解率 0.942 を出す。
///
/// データは Keras が配布する `mnist.npz`（NumPy の `.npy` を ZIP でまとめたもの）を、
/// 3 言語で共有して読む。F# には NumPy の読み込みが無いので、
/// **必要な形式（`|u1`・C 順）だけを読む小さなパーサ** をここに書く。
module GrokkingMlLib.Nb15ImageRecognition

open System.IO
open System.IO.Compression
open System.Text.RegularExpressions
open Accord.Neuro
open Accord.Neuro.Learning

/// 画像の一辺
[<Literal>]
let ImageSize = 28

/// 入力の次元。原著の `reshape(-1, 28*28)` と同じ
[<Literal>]
let InputDim = 784

/// 原著の隠れ層
let hiddenUnits = [| 128; 64 |]

/// クラス数（数字 0〜9）
[<Literal>]
let Classes = 10

/// 原著の学習回数
[<Literal>]
let OriginalEpochs = 10

/// 読み込んだ MNIST
type Mnist =
    { XTrain: float[][]
      YTrain: int[]
      XTest: float[][]
      YTest: int[] }

/// `.npy` から取り出した符号なし 8 ビットの配列
type NpyArray = { Shape: int[]; Data: byte[] }

let private shapePattern = Regex(@"'shape':\s*\(([^)]*)\)")

/// `.npy` を読む。原著の MNIST は `|u1`（符号なし 1 バイト）だけなのでそこだけ扱う
let parseNpy (bytes: byte[]) =
    if bytes.[0] <> 0x93uy || System.Text.Encoding.ASCII.GetString(bytes, 1, 5) <> "NUMPY" then
        failwith "npy のマジックナンバーが違います"

    // 版が 1 ならヘッダ長は 2 バイト、2 以降は 4 バイト（どちらもリトルエンディアン）
    let major = int bytes.[6]
    let lengthSize = if major = 1 then 2 else 4

    let headerLength =
        if lengthSize = 2 then
            int (System.BitConverter.ToUInt16(bytes, 8))
        else
            int (System.BitConverter.ToUInt32(bytes, 8))

    let header = System.Text.Encoding.ASCII.GetString(bytes, 8 + lengthSize, headerLength)

    if not (header.Contains "'descr': '|u1'") then
        failwithf "対応していない dtype です: %s" header

    let shape =
        shapePattern.Match(header).Groups.[1].Value.Split(',')
        |> Array.map (fun part -> part.Trim())
        |> Array.filter (fun part -> part <> "")
        |> Array.map int

    let offset = 8 + lengthSize + headerLength
    { Shape = shape; Data = bytes.[offset..] }

/// `mnist.npz`（ZIP）から 1 本の `.npy` を取り出す
let readNpz (name: string) =
    use archive = ZipFile.OpenRead(Datasets.path "mnist.npz")

    let entry =
        match archive.GetEntry name with
        | null -> failwithf "%s が mnist.npz にありません" name
        | entry -> entry

    use stream = entry.Open()
    use memory = new MemoryStream()
    stream.CopyTo memory
    parseNpy (memory.ToArray())

/// MNIST を読み込む。画素値は 0〜255 のまま返す
let loadMnist () =
    let images name =
        let array = readNpz name

        Array.init array.Shape.[0] (fun row -> Array.init InputDim (fun column -> float array.Data.[row * InputDim + column]))

    let labels name =
        let array = readNpz name
        Array.init array.Shape.[0] (fun i -> int array.Data.[i])

    { XTrain = images "x_train.npy"
      YTrain = labels "y_train.npy"
      XTest = images "x_test.npy"
      YTest = labels "y_test.npy" }

/// 学習した重みの総数。原著の `model.summary()` が出す 109,386 と突き合わせる
let parameterCount =
    (InputDim * hiddenUnits.[0] + hiddenUnits.[0])
    + (hiddenUnits.[0] * hiddenUnits.[1] + hiddenUnits.[1])
    + (hiddenUnits.[1] * Classes + Classes)

/// ラベルを one-hot に直す。Accord の出力層はクラスごとに 1 つのシグモイドになる
let private oneHot (label: int) =
    Array.init Classes (fun i -> if i = label then 1.0 else 0.0)

/// 原著と同じ形のネットワークを学習する。
///
/// 原著と違うところが 2 つある。どちらも **実測して必要だと分かった** ものである。
///
/// 1. **画素値を 0〜1 に直してから渡す。** 0〜255 のままだとシグモイドが飽和して
///    学習が進まない。
/// 2. **`NguyenWidrow(network).Randomize()` が必須。**
///    [#13](Nb13NeuralNetworkBoundary.fs) と同じで、呼ばないと重みが初期化されず
///    何エポック回しても正解率が動かない。エラーも警告も出ない。
///
/// 出力層はソフトマックスではなくシグモイドが 10 個並ぶ形になる。
/// 予測はそのうち最大のものを選ぶ（`np.argmax` と同じ）。
let fit (mnist: Mnist) (epochs: int) (sampleSize: int) (seed: int) =
    Accord.Math.Random.Generator.Seed <- System.Nullable seed

    let inputs = Array.init sampleSize (fun row -> mnist.XTrain.[row] |> Array.map (fun v -> v / 255.0))
    let outputs = Array.init sampleSize (fun row -> oneHot mnist.YTrain.[row])

    let network =
        ActivationNetwork(SigmoidFunction(), InputDim, hiddenUnits.[0], hiddenUnits.[1], Classes)

    NguyenWidrow(network).Randomize()

    let teacher = BackPropagationLearning(network)

    for _ in 1..epochs do
        teacher.RunEpoch(inputs, outputs) |> ignore

    network

/// 1 枚を予測する。出力 10 個のうち最大のものを選ぶ
let predict (network: ActivationNetwork) (image: float[]) =
    let scaled = image |> Array.map (fun v -> v / 255.0)
    let output = network.Compute scaled
    output |> Array.mapi (fun index value -> index, value) |> Array.maxBy snd |> fst

/// テストセットに対する正解率
let testAccuracy (network: ActivationNetwork) (mnist: Mnist) (sampleSize: int) =
    let correct =
        seq { 0 .. sampleSize - 1 }
        |> Seq.filter (fun row -> predict network mnist.XTest.[row] = mnist.YTest.[row])
        |> Seq.length

    float correct / float sampleSize

File: apps/grokking-ml-kotlin-lib/src/main/kotlin/lib/Nb15ImageRecognition.kt
package lib

import java.util.zip.ZipFile
import smile.base.mlp.Layer
import smile.base.mlp.OutputFunction
import smile.classification.MLP

/**
 * 原著ノートブック #15 `Chapter_10_Neural_Networks/Image_recognition.ipynb`。
 *
 * MNIST の手書き数字 7 万枚を、[#13][Nb13NeuralNetworkBoundary] と同じ形の
 * ネットワークで分類する。違いは入力が 784 次元（28 × 28 画素）で、
 * 出力が 10 クラスになること。
 *
 * **Kotlin に MNIST のローダは無い。** 原著の
 * `keras.datasets.mnist.load_data()` は `mnist.npz` を取ってくるだけなので、
 * 同じファイルを共有データセットから読み、`.npy` の中身を自前で解析する。
 */
object Nb15ImageRecognition {

    /** 画像の大きさ */
    const val IMAGE_SIZE = 28
    const val INPUT_DIM = IMAGE_SIZE * IMAGE_SIZE

    /** 原著のネットワークの形 */
    val HIDDEN_UNITS = intArrayOf(128, 64)
    const val CLASSES = 10

    /**
     * 原著の学習回数は 10 だが、Smile では時間がかかりすぎる。
     * 6 万枚 × 10 回は現実的でないので、既定は小さくしてある。
     */
    const val ORIGINAL_EPOCHS = 10

    /** MNIST の訓練セットとテストセット */
    class Mnist(
        val xTrain: Array<DoubleArray>,
        val yTrain: IntArray,
        val xTest: Array<DoubleArray>,
        val yTest: IntArray,
    )

    /**
     * NumPy の `.npy` 形式を読む。
     *
     * 形式は単純である。
     * - 6 バイトの magic `\x93NUMPY`
     * - メジャー・マイナー版が 1 バイトずつ
     * - ヘッダ長（版 1 なら 2 バイトのリトルエンディアン）
     * - Python の辞書リテラル（`{'descr': '|u1', 'fortran_order': False, 'shape': (60000, 28, 28), }`）
     * - 生のデータ
     *
     * MNIST は符号なし 8 ビット（`|u1`）なので、そのまま読める。
     */
    class NpyArray(val shape: IntArray, val data: ByteArray) {
        val size: Int get() = shape.fold(1) { acc, dimension -> acc * dimension }

        /** 符号なしとして解釈した値 */
        fun unsigned(index: Int): Int = data[index].toInt() and 0xFF
    }

    /** `.npy` の中身を解析する */
    fun parseNpy(bytes: ByteArray): NpyArray {
        require(bytes.size > 10) { "npy として短すぎます" }
        require(bytes[0] == 0x93.toByte() && String(bytes, 1, 5) == "NUMPY") {
            "npy の magic が合いません"
        }

        // 版 1 はヘッダ長が 2 バイト、版 2 以降は 4 バイト
        val major = bytes[6].toInt()
        val headerLengthBytes = if (major == 1) 2 else 4
        var headerLength = 0
        for (i in 0 until headerLengthBytes) {
            headerLength = headerLength or ((bytes[8 + i].toInt() and 0xFF) shl (8 * i))
        }
        val header = String(bytes, 8 + headerLengthBytes, headerLength)

        require("'descr': '|u1'" in header) { "符号なし 8 ビット以外は扱いません: $header" }

        val shape = SHAPE_PATTERN.find(header)!!.groupValues[1]
            .split(",")
            .map { it.trim() }
            .filter { it.isNotEmpty() }
            .map { it.toInt() }
            .toIntArray()

        val dataStart = 8 + headerLengthBytes + headerLength
        return NpyArray(shape, bytes.copyOfRange(dataStart, bytes.size))
    }

    private val SHAPE_PATTERN = Regex("""'shape':\s*\(([^)]*)\)""")

    /** `.npz`（zip）から指定した `.npy` を取り出して解析する */
    fun readNpz(name: String): NpyArray {
        ZipFile(Datasets.path("mnist.npz").toFile()).use { zip ->
            val entry = zip.getEntry(name) ?: error("$name が mnist.npz にありません")
            return parseNpy(zip.getInputStream(entry).readBytes())
        }
    }

    /**
     * MNIST を読み込む。
     *
     * 画像は 28 × 28 の 2 次元だが、[NpyArray] の時点で 1 次元に並んでいるので、
     * 784 ごとに切り出すだけで原著の `reshape(-1, 28*28)` と同じ形になる。
     */
    fun loadMnist(): Mnist {
        fun images(name: String): Array<DoubleArray> {
            val array = readNpz(name)
            val count = array.shape[0]
            return Array(count) { row ->
                DoubleArray(INPUT_DIM) { column -> array.unsigned(row * INPUT_DIM + column).toDouble() }
            }
        }

        fun labels(name: String): IntArray {
            val array = readNpz(name)
            return IntArray(array.shape[0]) { array.unsigned(it) }
        }

        return Mnist(images("x_train.npy"), labels("y_train.npy"), images("x_test.npy"), labels("y_test.npy"))
    }

    /** 原著の `batch_size` と同じ。**この値がそのまま学習の成否を決める**（[fit] 参照） */
    const val BATCH_SIZE = 10

    /** 学習率。Smile の既定（0.01）では 3 エポックでは足りないので上げてある */
    const val LEARNING_RATE = 0.1

    /**
     * 原著と同じ形のネットワークを学習する。
     *
     * 原著と違うところが 2 つある。どちらも **実測して必要だと分かった** ものである。
     *
     * 1. **画素値を 0〜1 に直してから渡す。** 原著は 0〜255 のまま渡しているが、
     *    [#14][Nb14HousePriceNetwork] と同じ理由で Smile では発散する。
     * 2. **[BATCH_SIZE] 件ずつに切って `update` を呼ぶ。**
     *    Smile の `MLP.update(x, y)` は渡した配列全体を 1 つのミニバッチとして扱う。
     *    訓練データをまるごと渡すと 1 エポックで重みが 1 回しか動かず、
     *    **正解率は当てずっぽうと同じ 0.09〜0.10 のまま** で、エラーも警告も出ない。
     *    10 件ずつに切ると、同じ 10000 件・3 エポックで 0.93 に届く。
     */
    fun fit(mnist: Mnist, epochs: Int, sampleSize: Int, seed: Int = 0): MLP {
        smile.math.MathEx.setSeed(seed.toLong())

        val x = Array(sampleSize) { row -> DoubleArray(INPUT_DIM) { mnist.xTrain[row][it] / 255.0 } }
        val y = IntArray(sampleSize) { mnist.yTrain[it] }

        val network = MLP(
            Layer.input(INPUT_DIM),
            Layer.rectifier(HIDDEN_UNITS[0]),
            Layer.rectifier(HIDDEN_UNITS[1]),
            Layer.mle(CLASSES, OutputFunction.SOFTMAX),
        )
        network.setLearningRate(smile.math.TimeFunction.constant(LEARNING_RATE))
        repeat(epochs) {
            var start = 0
            while (start + BATCH_SIZE <= sampleSize) {
                network.update(
                    Array(BATCH_SIZE) { x[start + it] },
                    IntArray(BATCH_SIZE) { y[start + it] },
                )
                start += BATCH_SIZE
            }
        }
        return network
    }

    /** 学習した重みの総数。原著の `model.summary()` が出す 109,386 と突き合わせる */
    fun parameterCount(): Int =
        (INPUT_DIM * HIDDEN_UNITS[0] + HIDDEN_UNITS[0]) +
            (HIDDEN_UNITS[0] * HIDDEN_UNITS[1] + HIDDEN_UNITS[1]) +
            (HIDDEN_UNITS[1] * CLASSES + CLASSES)

    /** テストセットに対する正解率 */
    fun testAccuracy(model: MLP, mnist: Mnist, sampleSize: Int = mnist.yTest.size): Double {
        val correct = (0 until sampleSize).count { row ->
            val scaled = DoubleArray(INPUT_DIM) { mnist.xTest[row][it] / 255.0 }
            model.predict(scaled) == mnist.yTest[row]
        }
        return correct.toDouble() / sampleSize
    }
}
